app/services/elasticsearch_handler.rb

```diff
@@ -0,0 +1,181 @@
+class ElasticsearchHandler
+  attr_reader :verbose
+  def initialize(verbose:false)
+    @verbose = verbose
+  end
+
+  def create_indices(client=Elasticsearch::Model.client)
+    drop_indices(client)
+    FolderFilesResponse.indexed_models.each do |indexed_model|
+      client.indices.create(
+        index: indexed_model.versioned_index_name,
+        body: {
+          settings: indexed_model.settings.to_hash,
+          mappings: indexed_model.mappings.to_hash
+        }
+      )
+
+      if client.indices.exists_alias? name: indexed_model.index_name
+        client.indices.get_alias(name: indexed_model.index_name).keys.each do |aliased_index|
+          $stderr.puts "consider deleting index #{aliased_index}" if @verbose
+          client.indices.delete_alias index: aliased_index, name: indexed_model.index_name
+        end
+      end
+      client.indices.put_alias index: indexed_model.versioned_index_name, name: indexed_model.index_name
+    end
+  end
+
+  def index_documents
+    batch_size = 500
+    FolderFilesResponse.indexed_models.each do |indexed_model|
+      indexed_model.paginates_per batch_size
+      (1 .. indexed_model.page.total_pages).each do |page_num|
+        current_batch = indexed_model
+          .page(page_num)
+          .map { |f|
+            { index: {
+              _index: indexed_model.versioned_index_name,
+              _type: f.__elasticsearch__.document_type,
+              _id: f.__elasticsearch__.id,
+              data: f.__elasticsearch__.as_indexed_json }
+            }
+        }
+        trys = 0
+        error_ids = []
+        while trys < 5
+          bulk_response = Elasticsearch::Model.client.bulk body: current_batch
+          if bulk_response["errors"]
+            trys += 1
+            error_ids = bulk_response["items"].select {|item|
+              item["index"]["status"] >= 400
+            }.map {|i|
+              i["index"]["_id"]
+            }
+            current_batch = current_batch.select {|b| error_ids.include? b[:index][:_id] }
+          else
+            trys = 5
+          end
+        end
+        if @verbose
+          unless error_ids.empty?
+            $stderr.puts "page #{page_num} Ids Not Loaded after #{trys} tries:"
+            $stderr.puts error_ids.join(',')
+          end
+          $stderr.print "+" * (current_batch.length - error_ids.length)
+        end
+      end
+    end
+  end
+
+  def drop_indices(client=Elasticsearch::Model.client)
+    client.indices.delete index: '_all'
+  end
+
+  def smart_reindex_indices
```

Comment:
This method is 42 lines long, with `if` statements nested 4 levels deep. It /should/ be refactored, but in the interest of time, we can pair tomorrow and walk through the code together.